project bootable/recovery/
diff --git a/Android.mk b/Android.mk
index f82c831..32452cb 100644
--- a/Android.mk
+++ b/Android.mk
@@ -103,6 +103,10 @@ ifeq ($(ENABLE_LOKI_RECOVERY),true)
   LOCAL_SRC_FILES += loki/loki_recovery.c
 endif
 
+ifeq ($(TARGET_RECOVERY_NO_RAINBOWS),true)
+  LOCAL_CFLAGS += -DNO_RAINBOWS
+endif
+
 # This binary is in the recovery ramdisk, which is otherwise a copy of root.
 # It gets copied there in config/Makefile.  LOCAL_MODULE_TAGS suppresses
 # a (redundant) copy of the binary in /system/bin for user builds.
diff --git a/recovery.c b/recovery.c
index 619049b..35b7cff 100644
--- a/recovery.c
+++ b/recovery.c
@@ -603,6 +603,7 @@ get_menu_selection(const char** headers, char** items, int menu_only,
             chosen_item = action;
         }
 
+#ifndef NO_RAINBOWS
         if (abs(selected - old_selected) > 1) {
             wrap_count++;
             if (wrap_count == 5) {
@@ -617,6 +618,7 @@ get_menu_selection(const char** headers, char** items, int menu_only,
                 }
             }
         }
+#endif
     }
 
     ui_end_menu();
