diff --git a/BoardConfig.mk b/BoardConfig.mk
index c27538e..2b57594 100644
--- a/BoardConfig.mk
+++ b/BoardConfig.mk
@@ -46,13 +46,13 @@ BOARD_WLAN_DEVICE := ath6kl
 # ATH6KL uses NL80211 driver
 WPA_SUPPLICANT_VERSION := VER_0_8_X
 BOARD_WPA_SUPPLICANT_DRIVER := NL80211
-BOARD_WPA_SUPPLICANT_PRIVATE_LIB := lib_driver_cmd_ath6kl
+# BOARD_WPA_SUPPLICANT_PRIVATE_LIB := lib_driver_cmd_ath6kl
 # Station/client mode
 WIFI_DRIVER_MODULE_NAME := "ath6kl"
 WIFI_DRIVER_MODULE_PATH := "/system/lib/modules/ath6kl.ko"
 # ATH6KL uses hostapd built from source
 BOARD_HOSTAPD_DRIVER := NL80211
-BOARD_HOSTAPD_PRIVATE_LIB := lib_driver_cmd_ath6kl
+# BOARD_HOSTAPD_PRIVATE_LIB := lib_driver_cmd_ath6kl
 # FW path parameters (todo: temp, should be ignored)
 WIFI_DRIVER_FW_PATH_STA := "sta"
 WIFI_DRIVER_FW_PATH_AP  := "ap"
diff --git a/init.tenderloin.rc b/init.tenderloin.rc
index aae71a5..5b56a07 100755
--- a/init.tenderloin.rc
+++ b/init.tenderloin.rc
@@ -164,7 +164,7 @@ on property:persist.service.adb.enable=1
 on property:persist.service.adb.enable=0
     stop adbd
 
-service wpa_supplicant /system/bin/wpa_supplicant -Dnl80211 -iwlan0 \
+service wpa_supplicant /system/bin/wpa_supplicant -dd -Dnl80211 -iwlan0 \
     -c/data/misc/wifi/wpa_supplicant.conf -e/data/misc/wifi/entropy.bin
     #   we will start as root and wpa_supplicant will switch to user wifi
     #   after setting up the capabilities required for WEXT
@@ -204,7 +204,12 @@ service iprenew_wlan0 /system/bin/dhcpcd -n
 #    disabled
 #    oneshot
 
-service hostapd_bin /system/bin/hostapd /data/misc/wifi/hostapd.conf
+service dhcpcd_bnep0 /system/bin/dhcpcd -BKLG
+    class main
+    disabled
+    oneshot
+
+service hostapd_bin /system/bin/hostapd -d /data/misc/wifi/hostapd.conf
     socket wpa_wlan0 dgram 660 wifi wifi
     disabled
     oneshot
