project hardware/atheros/wlan/
diff --git a/ath6kl/wpa_supplicant_8_lib/driver_cmd_nl80211.c b/ath6kl/wpa_supplicant_8_lib/driver_cmd_nl80211.c
index c106a73..9da428e 100644
--- a/ath6kl/wpa_supplicant_8_lib/driver_cmd_nl80211.c
+++ b/ath6kl/wpa_supplicant_8_lib/driver_cmd_nl80211.c
@@ -47,6 +47,8 @@ int wpa_driver_nl80211_driver_cmd(void *priv, char *cmd, char *buf,
 	android_wifi_priv_cmd priv_cmd;
 	int ret = 0;
 
+	wpa_printf(MSG_ERROR, "%s: CMD=%s", __func__, cmd);
+
 	if (os_strcasecmp(cmd, "STOP") == 0) {
 		linux_set_iface_flags(drv->global->ioctl_sock, bss->ifname, 0);
 		wpa_msg(drv->ctx, MSG_INFO, WPA_EVENT_DRIVER_STATE "STOPPED");
@@ -74,8 +76,10 @@ int wpa_driver_nl80211_driver_cmd(void *priv, char *cmd, char *buf,
 		return 0; /* Ignore it */
 	} else if(os_strncmp(cmd, "SETSUSPENDMODE ", 15) == 0) {
 		return 0; /* Ignore it */
+#if 1
 	} else if(os_strncmp(cmd, "WLS_BATCHING ", 13) == 0) {
 		return 0; /* Ignore it */
+#endif
 	} else if(os_strncmp(cmd, "BTCOEXSCAN-START", 16) == 0) {
 		return 0; /* Ignore it */
 	} else if(os_strncmp(cmd, "BTCOEXSCAN-STOP", 15) == 0) {
