diff --git a/arch/arm/mach-msm/board-msm8x60.c b/arch/arm/mach-msm/board-msm8x60.c
index de5738c..d7ae644 100644
--- a/arch/arm/mach-msm/board-msm8x60.c
+++ b/arch/arm/mach-msm/board-msm8x60.c
@@ -6298,6 +6298,17 @@ static struct gpio_keys_button topaz_wifi_gpio_keys_buttons[] = {
 		.debounce_interval = 50,
 		.wakeup		= 1
 	},
+#if 0 && defined(CONFIG_USB_PEHCI_HCD)
+	{
+		.code           = KEY_UIM_INSERT,
+		.gpio           = (GPIO_3G_UIM_CD_N),
+		.desc           = "mdmuim",
+		.active_low     = 1,
+		.type		= EV_KEY,
+		.debounce_interval = 500,
+		.wakeup		= 1
+	},
+#endif
 };
 
 static struct gpio_keys_button topaz_3g_gpio_keys_buttons[] = {
@@ -6325,17 +6336,6 @@ static struct gpio_keys_button topaz_3g_gpio_keys_buttons[] = {
 		.type		= EV_KEY,
 		.wakeup		= 0
 	},
-#ifdef CONFIG_USB_PEHCI_HCD
-	{
-		.code           = KEY_UIM_INSERT,
-		.gpio           = (GPIO_3G_UIM_CD_N),
-		.desc           = "mdmuim",
-		.active_low     = 1,
-		.type		= EV_KEY,
-		.debounce_interval = 500,
-		.wakeup		= 1
-	},
-#endif
 };
 
 static struct gpio_keys_platform_data topaz_3g_gpio_keys_data = {
