diff --git a/emailcommon/src/com/android/emailcommon/utility/AttachmentUtilities.java b/emailcommon/src/com/android/emailcommon/utility/AttachmentUtilities.java
index 969e542..80d809d 100644
--- a/emailcommon/src/com/android/emailcommon/utility/AttachmentUtilities.java
+++ b/emailcommon/src/com/android/emailcommon/utility/AttachmentUtilities.java
@@ -114,7 +114,7 @@ public class AttachmentUtilities {
         "mst", "pif", "scr", "sct", "shb", "sys", "vb", "vbe",
         "vbs", "vxd", "wsc", "wsf", "wsh",
         // File types of common compression/container formats (again, to avoid malware)
-        "zip", "gz", "z", "tar", "tgz", "bz2",
+        /* "zip", "gz", "z", "tar", "tgz", "bz2", */
     };
     /**
      * Filename extensions of attachments that can be installed.
