diff --git a/config/common.mk b/config/common.mk
index b83b994..f2f68e4 100644
--- a/config/common.mk
+++ b/config/common.mk
@@ -228,39 +228,42 @@ ifndef CM_BUILDTYPE
         # Starting with "CM_" is optional
         RELEASE_TYPE := $(shell echo $(RELEASE_TYPE) | sed -e 's|^CM_||g')
         CM_BUILDTYPE := $(RELEASE_TYPE)
+        CM_EXTRAVERSION_DATESTAMP := TRUE
+        CM_EXTRAVERSION_TIMESTAMP_UTC := TRUE
     endif
 endif
 
-# Filter out random types, so it'll reset to UNOFFICIAL
-ifeq ($(filter RELEASE NIGHTLY SNAPSHOT EXPERIMENTAL,$(CM_BUILDTYPE)),)
-    CM_BUILDTYPE :=
+# If CM_BUILDTYPE is not defined, set to UNOFFICIAL
+ifndef CM_BUILDTYPE
+    CM_BUILDTYPE := UNOFFICIAL
+    CM_EXTRAVERSION :=
 endif
 
-ifdef CM_BUILDTYPE
-    ifneq ($(CM_BUILDTYPE), SNAPSHOT)
-        ifdef CM_EXTRAVERSION
-            # Force build type to EXPERIMENTAL
-            CM_BUILDTYPE := EXPERIMENTAL
-            # Remove leading dash from CM_EXTRAVERSION
-            CM_EXTRAVERSION := $(shell echo $(CM_EXTRAVERSION) | sed 's/-//')
-            # Add leading dash to CM_EXTRAVERSION
-            CM_EXTRAVERSION := -$(CM_EXTRAVERSION)
+ifndef CM_EXTRAVERSION
+    ifdef CM_EXTRAVERSION_TIMESTAMP
+        ifdef CM_EXTRAVERSION_TIMESTAMP_UTC
+            CM_EXTRAVERSION_TIMESTAMP := $(shell date -u +%Y%m%d_%H%M)
+        else
+            CM_EXTRAVERSION_TIMESTAMP := $(shell date +%Y%m%d_%H%M)
         endif
-    else
-        ifndef CM_EXTRAVERSION
-            # Force build type to EXPERIMENTAL, SNAPSHOT mandates a tag
-            CM_BUILDTYPE := EXPERIMENTAL
+    else ifdef CM_EXTRAVERSION_DATESTAMP
+        ifdef CM_EXTRAVERSION_TIMESTAMP_UTC
+            CM_EXTRAVERSION_TIMESTAMP := $(shell date -u +%Y%m%d)
         else
-            # Remove leading dash from CM_EXTRAVERSION
-            CM_EXTRAVERSION := $(shell echo $(CM_EXTRAVERSION) | sed 's/-//')
-            # Add leading dash to CM_EXTRAVERSION
-            CM_EXTRAVERSION := -$(CM_EXTRAVERSION)
+            CM_EXTRAVERSION_TIMESTAMP := $(shell date +%Y%m%d)
+        endif
+    endif
+    ifdef CM_EXTRAVERSION_TAG
+        ifdef CM_EXTRAVERSION_TIMESTAMP
+            CM_EXTRAVERSION := $(CM_EXTRAVERSION_TIMESTAMP)-$(CM_EXTRAVERSION_TAG)
+        else
+            CM_EXTRAVERSION := $(CM_EXTRAVERSION_TAG)
+        endif
+    else
+        ifdef CM_EXTRAVERSION_TIMESTAMP
+            CM_EXTRAVERSION := $(CM_EXTRAVERSION_TIMESTAMP)
         endif
     endif
-else
-    # If CM_BUILDTYPE is not defined, set to UNOFFICIAL
-    CM_BUILDTYPE := UNOFFICIAL
-    CM_EXTRAVERSION :=
 endif
 
 ifeq ($(CM_BUILDTYPE), UNOFFICIAL)
@@ -280,10 +283,10 @@ ifeq ($(CM_BUILDTYPE), RELEASE)
         endif
     endif
 else
-    ifeq ($(PRODUCT_VERSION_MINOR),0)
-        CM_VERSION := $(PRODUCT_VERSION_MAJOR)-$(shell date -u +%Y%m%d)-$(CM_BUILDTYPE)$(CM_EXTRAVERSION)-$(CM_BUILD)
+    ifdef CM_EXTRAVERSION
+        CM_VERSION := $(PRODUCT_VERSION_MAJOR)-$(CM_BUILDTYPE)-$(CM_EXTRAVERSION)-$(CM_BUILD)
     else
-        CM_VERSION := $(PRODUCT_VERSION_MAJOR).$(PRODUCT_VERSION_MINOR)-$(shell date -u +%Y%m%d)-$(CM_BUILDTYPE)$(CM_EXTRAVERSION)-$(CM_BUILD)
+        CM_VERSION := $(PRODUCT_VERSION_MAJOR)-$(CM_BUILDTYPE)-$(CM_BUILD)
     endif
 endif
 
